modules/layers/src/path-layer/tesselator.js

```diff
@@ -0,0 +1,159 @@
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+import {experimental} from '@deck.gl/core';
+const {fillArray} = experimental;
+
+export class TypedArrayManager {
+  constructor({overAlloc = 1} = {}) {
+    this.overAlloc = overAlloc;
+  }
+
+  allocate(Type, size) {
+    if (Type === Uint16Array && size > 65535) {
+      throw new Error('Vertex count exceeds browser index limit');
+    }
+    // TODO - check if available in pool
+    return new Type(size);
+  }
+
+  reallocate(typedArray, newSize, copy = false) {
+    if (newSize <= typedArray.length) {
+      return typedArray;
+    }
+
+    // Allocate at least one element to ensure a valid buffer
+    const allocSize = Math.max(Math.ceil(newSize * this.overAlloc), 1);
+    const newArray = this.allocate(typedArray.constructor, allocSize);
+
+    if (copy) {
+      newArray.set(typedArray);
+    }
+
+    this.release(typedArray);
+    return newArray;
+  }
+
+  release(typedArray) {
+    // TODO - add to pool
+    // logFunctions.onUpdate({
+    //   level: LOG_DETAIL_PRIORITY,
+    //   message: `${attributeName} allocated ${allocCount}`,
+    //   id: this.id
+    // });
+  }
+}
+
+export default class Tesselator {
+  constructor(opts = {}) {
+    const {attributes = {}} = opts;
+
+    this.typedArrayManager = new TypedArrayManager();
+    this.bufferLayout = null;
+    this.instanceCount = 0;
+    this.attributes = {};
+    this._attributeDefs = attributes;
+
+    this.updateGeometry(opts);
+
+    Object.seal(this);
+  }
+
+  /* Public methods */
+  updateGeometry({data, getGeometry, fp64}) {
+    this.data = data;
+    this.getGeometry = getGeometry;
+    this.fp64 = fp64;
+    this._rebuildGeometry();
+  }
+
+  /* Subclass interface */
+
+  // Fill the typed arrays for geometry attributes
+  updateGeometryAttributes() {
+    throw new Error('Not implemented');
+  }
+
+  // Returns the number of vertices in a geometry
+  countInstancesInGeometry(geometry) {
```

Comment:
Hmm.. For the general case, we need to get both number of indices and vertices for each sub-geometry. Since counting them separately likely requires double work, this function should return two values, where indices could be `null` or `undefined`.

Since instances are just a special case of vertices, not sure it needs a separate method.

Reply:
In path layer, each segment is rendered as an instance, so the count returned by this function is actually `vertexCount - 1`. I agree it's not an ideal name since not all users of this class are instanced.

Reply:
Hmm, not sure I follow if that means we need a separate callback, but no matter for now.

Maybe we can consolidate the two callbacks, or maybe need a separate tesselator callback for instanced layers. 

The key to my thinking is that the method I outlined should be the core of each `Tesselator` subclass.

Reply:
My thinking is that in the final version, the tesselator first loops over all objects and gets their sizes and then loops over all objects and populates their geometries, and these are the two main functions that subclasses should override. 

I struggled quite a bit to find names for them that I liked, but at least they should be named as a pair, maybe something like:

* `getObjectGeometrySize({object})`
* `updateObjectGeometry({object, start, count, indexStart, indexCount, ...})`